src/frontend/flight/sync.rs

```diff
@@ -371,42 +400,170 @@ impl SeafowlDataSyncManager {
 
-        // TODO: Filter away deletes for now
-        let filter_plan: Arc<dyn ExecutionPlan> = Arc::new(FilterExec::try_new(
-            col(SEAFOWL_SYNC_DATA_UD_FLAG, schema.as_ref())?,
-            mem_plan,
-        )?);
+        let sync_df = self.context.inner.read_batches(data)?;
+
+        // First remove all deleted rows
+        let join_cols = pk_columns.iter().map(|pk| pk.as_str()).collect::<Vec<_>>();
+        let input_df = input_df.join(
+            sync_df
+                .clone()
+                .filter(is_false(col(SEAFOWL_SYNC_DATA_UD_FLAG)))?,
+            JoinType::LeftAnti,
+            &join_cols,
+            &join_cols,
+            None,
+        )?;
 
         // Normalize the schema, by ordering columns according to the full table schema and
-        // projecting any missing columns as NULLs.
+        // projecting the sync data accordingly.
         let projection = full_schema
             .all_fields()
             .iter()
             .map(|f| {
                 let name = f.name();
-                if schema.column_with_name(name).is_some() {
-                    Ok((col(name, schema.as_ref())?, name.to_string()))
+                Ok(when(
+                    // If the row is not present in the sync data, project the existing row
+                    is_null(col(SEAFOWL_SYNC_DATA_UD_FLAG)),
+                    Expr::Column(Column::new(Some(SYNC_REF), name)),
+                )
+                .otherwise(if schema.column_with_name(name).is_some() {
+                    // If the row and column are present in the sync project its value
+                    Expr::Column(Column::new(Some(UNNAMED_TABLE), name))
                 } else {
-                    Ok((
+                    when(
+                        is_null(col(SYNC_JOIN_COLUMN)),
+                        // Neither the old row nor the sync column exist, project a NULL
                         lit(ScalarValue::Null.cast_to(f.data_type())?),
-                        name.to_string(),
-                    ))
-                }
+                    )
+                    .otherwise(
+                        // Sync column doesn't exist but an old row does, project the old row
+                        Expr::Column(Column::new(Some(SYNC_REF), name)),
+                    )?
+                })?
+                .alias_qualified(Some(SYNC_REF), name))
             })
             .collect::<Result<_>>()?;
-        let proj_plan = Arc::new(ProjectionExec::try_new(projection, filter_plan)?);
 
-        Ok(Arc::new(UnionExec::new(vec![input_plan, proj_plan])))
+        let input_df = input_df
+            .with_column(SYNC_JOIN_COLUMN, lit(true))?
+            .join(
+                sync_df.filter(col(SEAFOWL_SYNC_DATA_UD_FLAG))?,
+                JoinType::Full,
+                &join_cols,
+                &join_cols,
+                None,
+            )?
+            .select(projection)?;
+
+        Ok(input_df)
```

Comment:
Will this write an empty partition if the entire partition gets cleaned out by a sync?

Reply:
It shouldn't write an empty partition but it may write an empty commit (without adds/removes). 

We could do some pre-exec aggregation later on to e.g. see how many rows would be produced, and if it's zero and there were no removes just skip the commit?

Reply:
Oh yeah I see, since this goes into `plan_to_object_store`. Let's test for what happens when an entire write is just deletes (later, after we've settled on the protocol).